### ISSUE TITLE
Investigate and remove HACK for native declarations

### ISSUE THREAD

**Issue by quintesse** [ceylon/ceylon-js#569]
@chochos could you maybe take a look at the hack I implemented to make JS compilation of native declarations work: https://github.com/ceylon/ceylon-spec/blob/master/src/com/redhat/ceylon/compiler/typechecker/analyzer/ExpressionVisitor.java#L10413

The typechecker looks both in the code and in the model loader to see if it can find the native implementation. On the JVM side that includes things implemented in Java, but possibly the JS has no way of adding those native files to the model? I'm not sure how it all works to be honest.

Just creating this issue so we don't forget to get rid of that hack somehow.

**Comment by chochos** [comment]
There are two kinds of native code in JS:

* Ceylon declarations annotated `native` can have their implementation done in JS, just a file with the JS code for that method/attribute/type/constructor/whatever. Since the declaration is in Ceylon, it's added to the model.
* Native files that are just copied to the resulting module. There is simply no way to know what's in there; the compiler simply copies the contents of the js file to the resulting module.

**Comment by quintesse** [comment]
Indeed, but the thing is how can the typechecker know about native implementations of the first kind having only access to the `JsModuleManager`? On the JVM side the `ModelLoader` takes care of that, but does the JS side do something similar? *Can* it do something similar?

The thing is that the typechecker needs to know whether the following code is illegal or not:

```ceylon
native shared void test();

shared void run() {
    test();  // If no implementation exists we need an error here
}
```

**Comment by quintesse** [comment]
I now made the hacks to be able to compile native JS code without the typechecker complaining a little less horrible (@6b92947 vs ceylon/ceylon-spec@0a75ee2c00004de8e9fe823147051d739d2cdcfc) but it's still something that has to have a proper solution.

Today I sent this to the mailing list as a possible way to tackle it:

```
So, due to the changes made in the typechecker for the native stuff it is now capable of
detecting missing native implementations.
For the JVM backend even when written manually in Java. The JVM's model loader handles
that just fine.

The problem I'm having is that the JS backend can't do the same. Its model loader
doesn't know about the .js files that you can put next to the .ceylon files which will then
be automatically "stitched" into the resulting JS output. And even if it did there's just not
enough information in those files to know what to do with them.
Right now when the JS compiler encounters a "native" declaration it will look for a file with
that name and if it exists it includes ("stitches") it. But it does this *after* the typechecker
has done all its work which is too late for it to be able to detect missing native implementations.

So to be able to make this work I was thinking about introducing a special Visitor for the JS
backend that visits all the native declarations, looks if an associated .js file exists and then
adds an overload (a copy of the header) for it so later typechecker phases can do their work
properly.

For this to function properly the visitor would need to run just *after* the declarations have
been scanned (PhasedUnit.scanDeclarations()). Thing is that I have no clear idea what the
proper way would be for a backend compiler to insert a Visitor of its own into the typechecker's
 `executePhases()`.
```

Gavin then updated the `TypeChecker` slightly to make it easier to override its functionality.

So basically I was thinking about a simple override of the `TypeCheckerBuilder` (named `JsTypeCheckerBuilder`) which creates a `JsTypeChecker` instead of a plain `TypeChecker`.
The `JsTypeChecker` would then re-implement `executePhases()` and add a `NativeDeclarationVisitor` phase just after the loop where `scanDeclarations()` is called.

**Comment by chochos** [comment]
@quintesse I guess we can close this now?

**Comment by quintesse** [comment]
Not sure actually, but so much has changed that I think this issue as it stands is of little use, so I'll close it.
